### PR TITLE
JDK-8317299: safepoint scalarization doesn't keep track of the depth of the JVM state

### PR DESCRIPTION
**Pull request by dafedafe**
# Issue

The origin of the problem is tied to the fact that, when C2 optimizes vector boxes, it performs safepoint object scalarization before late inlining.
This can lead to situations in which scalarization adds scalarized values to the JVM state and late inlining of further methods adds further JVM state entries on top for each inlined method.
With the example of the reported bug (_TestIntrinsicBailOut.java_) we get to a situation like this:
```
...
bc: JVMS depth=6 loc=20 stk=23 arg=23 mon=23 scalar=23 end=23 mondepth=0 sp=0 bci=1 reexecute=false method=virtual jobject jdk.incubator.vector.ByteVector.rearrangeTemplate(jobject, jobject)
bc: JVMS depth=7 loc=23 stk=27 arg=27 mon=27 scalar=27 end=27 mondepth=0 sp=0 bci=36 reexecute=false method=virtual jobject jdk.incubator.vector.AbstractShuffle.checkIndexes()
bc: JVMS depth=8 loc=27 stk=28 arg=28 mon=28 scalar=28 end=28 mondepth=0 sp=0 bci=1 reexecute=false method=virtual jobject jdk.incubator.vector.AbstractShuffle.reorder()
bc: JVMS depth=9 loc=28 stk=29 arg=29 mon=29 scalar=29 end=31 mondepth=0 sp=0 bci=1 reexecute=false method=virtual jobject jdk.internal.vm.vector.VectorSupport$VectorPayload.getPayload()
bc: JVMS depth=10 loc=31 stk=32 arg=32 mon=32 scalar=32 end=32 mondepth=0 sp=0 bci=3 reexecute=false method=static jobject jdk.internal.vm.vector.VectorSupport.maybeRebox(jobject)
bc: JVMS depth=11 loc=32 stk=33 arg=33 mon=33 scalar=33 end=33 mondepth=0 sp=0 bci=1 reexecute=false method=virtual void jdk.internal.misc.Unsafe.loadFence()
```
`JVMS depth=9` shows 2 scalars but 2 further inlines added 2 more JVM states (with no scalars).

The corresponding node looks like this:
image

To keep track of its scalarized inputs, `SafePointScalarObjectNode` keeps a field `_first_index`, which is supposed to be  "relative to the last (youngest) jvms->_scloff"...
https://github.com/openjdk/jdk/blob/c5e72450966ad50d57a8d22e9d634bfcb319aee9/src/hotspot/share/opto/callnode.hpp#L509-L511
but if there are late inlined methods, this field is going to be relative to the JVM state at the depth before inlining happened (e.g. depth=9 in the example) and not relative to the youngest depth.

# Solution

In order to keep track of the right depth a `_depth` field is added to `SafePointScalarObjectNode`, which refers to the depth of the JVM state the `_first_index` field refers to. The method `uint first_index(JVMState* jvms)` is adapted accordingly.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317299](https://bugs.openjdk.org/browse/JDK-8317299): safepoint scalarization doesn't keep track of the depth of the JVM state (**Bug** - P3)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [c5f693eb](https://git.openjdk.org/jdk/pull/17500/files/c5f693eba019ae59c5bc6fdddecfffe3d1dc4b4a)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17500/head:pull/17500` \
`$ git checkout pull/17500`

Update a local copy of the PR: \
`$ git checkout pull/17500` \
`$ git pull https://git.openjdk.org/jdk.git pull/17500/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17500`

View PR using the GUI difftool: \
`$ git pr show -t 17500`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17500.diff">https://git.openjdk.org/jdk/pull/17500.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17500#issuecomment-1907737886)

### CODE REVIEW

**Review by TobiHartmann — COMMENTED** [review]
Great work, Damon. When we discussed this, I always wondered why we don't hit the same issue in Valhalla, where we perform even more aggressive scalarization during IGVN. Turns out that I postponed scalarization to after inlining, to work around that exact problem:
https://github.com/openjdk/valhalla/blob/c48006dfc05bb0c41ab9ae55ead226356259c46d/src/hotspot/share/opto/compile.cpp#L2002

With your fix, we can remove that limitation. I filed [JDK-8324605](https://bugs.openjdk.org/browse/JDK-8324605) for this.

**Comment by vnkozlov** [comment]
Interesting. Could it resolve the issue [JDK-8276112](https://bugs.openjdk.org/browse/JDK-8276112) so we can REDO [DK-8276998](https://bugs.openjdk.org/browse/JDK-8276998)?

**Comment by TobiHartmann** [comment]
Good catch, Vladimir. I completely forgot about [JDK-8276112](https://bugs.openjdk.org/browse/JDK-8276112) but from re-reading my old analysis in https://github.com/openjdk/jdk/pull/6333, it's most likely the exact same issue.

File: src/hotspot/share/opto/callnode.hpp [hunk 1]
```diff
@@ -508,7 +508,7 @@ class SafePointNode : public MultiNode {
 class SafePointScalarObjectNode: public TypeNode {
   uint _first_index;              // First input edge relative index of a SafePoint node where
                                   // states of the scalarized object fields are collected.
-                                  // It is relative to the last (youngest) jvms->_scloff.
+  uint _depth;                    // Depth of the JVM state the first index field refers to
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/callnode.hpp:511]:
```suggestion
  uint _depth;                    // Depth of the JVM state the _first_index field refers to
```

**Review by TobiHartmann — APPROVED** [review]
The fix looks good to me.

**Comment by dafedafe** [comment]
> Interesting. Could it resolve the issue [JDK-8276112](https://bugs.openjdk.org/browse/JDK-8276112) so we can REDO [DK-8276998](https://bugs.openjdk.org/browse/JDK-8276998)?

Thanks for pointing this out @vnkozlov. I've tried reapplying [JDK-8261137](https://bugs.openjdk.org/browse/JDK-8261137) to reproduce the [JDK-8276998](https://bugs.openjdk.org/browse/JDK-8276998) issue before applying this change but couldn't reproduce it anymore. I'll mention this issue in the REDO anyway.

**Comment by dafedafe** [comment]
@TobiHartmann @iwanowww thanks a lot for reviewing!

**Pull request closed without merging**